### DISCUSSION TITLE
MXPredReshape always fails (return -1) - MXNet 1.3.0 C-API

### DISCUSSION THREAD

**Discussion by johnbroughton2017** [apache/incubator-mxnet#12814] [Q&A]
Hi friends,

I am having some problem with MXPredReshape(). I tried to reshape the input to a larger size but the function always returns -1. The code is given below. I am not sure what went wrong since there is not error hints. Any suggestion is highly appreciated. Thanks.

-- J

``

        // create Predictor
        mx_uint num_input_nodes = 1;  // 1 for feedforward
        const char* input_key[1] = {"data"};
        const char** input_keys = input_key;
        const mx_uint input_shape_indptr[2] = { 0, 4 };
        const mx_uint input_shape_data[4] = { batch_size, input_depth, input_height, input_width };
        MXPredCreatePartialOut( (const char*)model_buf.GetBuffer(),
                                (const char*)param_buf.GetBuffer(),
                                static_cast<size_t >(param_buf.GetLength()),
                                device_type,
                                device_id,
                                num_input_nodes,
                                input_keys,
                                input_shape_indptr,
                                input_shape_data,
                                0,
                                NULL, 
                                &pred_hnd);

        if (!pred_hnd) {
            LOG(ERROR) << "Fail to create MXNet predictor";
            return false;
        }
        
        PredictorHandle reshaped_pred_hnd = NULL;
        const mx_uint new_input_shape_data[4] = { batch_size, input_depth, 512, 512 };
        int ret = MXPredReshape(num_input_nodes,
                                input_keys,
                                input_shape_indptr,
                                new_input_shape_data,
                                pred_hnd,
                                &reshaped_pred_hnd);
``

**Comment by piyushghai** [comment]
@johnbroughton2017 Thank you for posting your question.
You can also post this question on MXNet discussion forum (https://discuss.mxnet.io), where it will get a wider audience and allow other to learn as well.

@mxnet-label-bot [C++, Question]

**Comment by johnbroughton2017** [comment]
done! thank you @piyushghai

**Comment by piyushghai** [comment]
@johnbroughton2017 Looking at the code snippet you pasted, 
Specifically : ```const mx_uint input_shape_data[4] = { batch_size, input_depth, input_height, input_width };```

What are the values of ```input_height``` and ```input_width``` ? 

Because you are reshaping it to : 
```const mx_uint new_input_shape_data[4] = { batch_size, input_depth, 512, 512 };```

So for the reshape to work, the product of : 
```batch_size * input_depth * input_height * input_width``` should be equal to ```batch_size * input_depth * 512 * 512 ```

**Comment by marcoabreu** [comment]
@piyushghai I have corrected the label assignment. This issue is about the C API which is part of our internal API design. 

@johnbroughton2017 Please note that the C-API is not intended for external usage and thus not governed by our semantic versioning. The API might change at any point in time, so please consider this in case you build a critical service on top of it.

**Comment by johnbroughton2017** [comment]
Thanks for your feedback @marcoabreu 

I am curious if mxnet will provide standard C APIs to developers soon. My application needs to run in realtime so our code is generally in C/C++. Thanks.

**Comment by marcoabreu** [comment]
Please have a look at https://github.com/apache/incubator-mxnet/tree/master/cpp-package

The cpp package is our user-facing frontend API that's written in c++. Does that suite your use case?

**Comment by lupesko** [comment]
@johnbroughton2017 as @marcoabreu pointed above, MXNet does have a C++ API you can use.
Feel free to check it out and close this issue if you have no more questions. Thanks!
